Final task Sprint 8 Changes: - added start date for Task, SubTask and Epic - added logic for epic start time, duration calculation - added logic for time save/parsing to/from the file - added saving of prioritized task by start date - added check of overlapped tasks - replaced for loop to stream API

imposmas opened the pull request.

imposmas committed:
Final task Sprint 8
Changes:
- added start date for Task, SubTask and Epic
- added logic for epic start time, duration calculation
- added logic for time save/parsing to/from the file
- added saving of prioritized task by start date
- added check of overlapped tasks
- replaced for loop to stream API

imposmas committed:
checkstyle correction

VadimZharkov commented on src/ru/yandex/practicum/kanban/generics/tasks/Task.java:
```diff
@@ -11,6 +12,8 @@ public class Task {
     protected TaskStatus status;
     protected int id;
+    protected LocalDateTime startTime;
+    protected int taskDuration;
```
вместо int лучше Duration

VadimZharkov commented on src/ru/yandex/practicum/kanban/managers/InMemoryTaskManager.java:
```diff
@@ -132,7 +121,10 @@ public Task findTaskById(int id) {
+        Task taskNew = new Task(id, task);
+        overlapWithExistingTasksCheck(taskNew);
+        tasks.put(id, taskNew);
+        prioritizedTasks.add(tasks.get(id));
```
Можно prioritizedTasks.add(taskNew); без дополнительного обращения к Map, будет чуть быстрей. Для нашего проекта не принципиально, но в нагруженных проектах по совокупности может сыграть свою роль в быстродействии.

VadimZharkov commented on src/ru/yandex/practicum/kanban/managers/InMemoryTaskManager.java:
```diff
@@ -249,11 +262,31 @@ private void calculateEpicStatus(Epic epic) {
 
-    public boolean isAllSubTasksNew(ArrayList<SubTask> subTasks) {
+    protected void calculateEpicStartTimeAndDuration(Epic epic) {
+        List<SubTask> subTasksPerEpic = new ArrayList<>(subTasks.values().stream()
```
У эпика есть список идентификаторов его подзадач. Эффективней взять его, чем перебирать все подзадачи.

VadimZharkov reviewed (changes requested):
Хорошая работа, логика реализована верно. Можно немного оптимизировать, см. комменты.

imposmas committed:
1. int --> Duration for taskDuration
2. corrected subtasks finding for epic
3. corrected adding task to prioritized tasks
